### DISCUSSION TITLE
Assignment - Create the class diagram for BookMyShow.

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#19] [General]
What?
Identify the various entities, their attributes, behaviour and relationships with other entities.

Why?
1. Helps identifying and documenting low level implementation
2. Removes uncertainty from actual implementation

How?
Fill out the worksheet [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/06-design-book-my-show-worksheet.md#class-diagrams).

From 10:00 - 10:15.

**Comment by MRKarthick** [comment]
![Uploading Screenshot 2023-01-20 at 5.34.46 AM.png…]()

**Reply by MRKarthick** [reply]
```
<Location>
Name
Logo
Type (Metropolitan, village) (used for determining cost)

<Theatre>
Name
Address
Rating
LocationId

<Screen>
Name
TheatreId
NoOfSeats
isAC
isDTS
Array of Array of Seats (Might not be populated at start)

<Seats>
Name
Cost
Type<Front Seats, Side Seats, Balcony Seats, Premium Location Seats, Couple Seats>

<Movie>
Name
CastDetails
BaseCost
Rating

<Ticket>
UserId
Name
LocationId
TheatreId
ScreenId
SeatId
Cost
CouponCost InvoiceId

<Show>
ScreenId
MovieId
Date 
Timings

<Coupon>

<Invoice>

<Location> 1->m <Theatre>
<Theatre> 1->m <Screen>
<Screen> 1->m <seats>
<Movie> 1->m <Show>
<Screen> 1 -> m <Show>

```

**Reply by kanmaytacker** [reply]
Few things to note -
1. You might need a couple of more entities like City and Payment.
2. In this design, once I book a seat for a show that seat is booked for all the shows. Try to think a way around it.

**Comment by illusionrs** [comment]
[https://drive.google.com/file/d/1SQx-_aJGAX26sF9siBT3mWe6C5C9VvpY/view?usp=sharing](url)

**Comment by sahilshaikh1634** [comment]
User - name,email,location
City- CityName,Theaters[],Location
Theaters - Name,Location,Screens[]
Screen - ScreenNo, Type, Movie, Time, Seats[]
Seats - SeatNo , Type, UserId, Status, Price
Ticket - Userid, No. of seats,Amount
Payment - TicketId,  Amout ,Type , Status.

**Comment by kishor-a** [comment]
Class Theatre
 -name
 -address
 -movieHalls
 -gates

Class MovieHall
 -number
 -screenSize
 -seats
 
Class Seat
 -number
 -type

Enum SeatType
 SILVER
 GOLD 
 PLATINUM

Movie
 -name
 -duration
 -cast
 -description
 -rating
 -languages

Class Show
 -time
 -movieHall
 -ShowStatus
 -prize
 +book()
 +cancel()

ENUM ShowStatus
 -NOT_STARTED
 -IN_PROGRESS
 -ENDED
 -CANCELLED

Class Ticket
 -show
 -Status
 -userId
 -seatNumber
 -date

ENUM TicketStatus
 PENDING
 BOOKED
 CANCELLED

User
 -id
 -name
 -age
 -contactNumber
 -email
 -UserRole
 +updateProfile()

ENUM UserRole
 ADMIN
 CUSTOMER
 THEATRE_OWNER
 CRITIC

**Reply by HarshitSingh96** [reply]
![classdiagBMS](https://user-images.githubusercontent.com/23292483/213504227-252e6421-b231-4765-9d5c-bfaf8874d45e.jpeg)

**Comment by ajayv1** [comment]
![WhatsApp Image 2023-01-19 at 10 11 43 PM](https://user-images.githubusercontent.com/11134190/213504028-fbd8f63d-0f89-4db1-87da-b5082db89791.jpeg)

**Comment by redpulse96** [comment]
Class User
- mobileNo
- bookingId

Class LocationDetails
- city
- branch

Class Cinema
- locationDetails
- noOfScreens
- movies: Movie[]
- shows: Show[]

Class Movie
- runTime
- showTime
- screenNo
- basePrice

Class Show
- movie: Movie
- baseTicketPrice
- noOfSeats
- noOfSeatsBooked

Class Seat
- isBooked
- bookingId
- showId
- ticketId

Class Ticket
- showId
- bookingId
- price
- seatId

**Reply by kanmaytacker** [reply]
Few things to note -
1. You might need a couple of more entities like City, Invoice, Payment.
2. In this design, once I book a seat for a show that seat is booked for all the shows. Try to think a way around it.

**Comment by Mariselvam** [comment]
@kanmaytacker   could you please check the below design and comment on it ..?

![bookmyshow-class-diagram](https://user-images.githubusercontent.com/1678425/213859846-52b6c2bf-9350-4711-94d0-9f0b00969a36.png)

**Comment by GouravSinghChouhan** [comment]
file:///home/gourav/Pictures/Screenshot%20from%202023-01-26%2022-37-10.png

**Comment by GouravSinghChouhan** [comment]
![Screenshot from 2023-01-26 22-37-10](https://user-images.githubusercontent.com/45309371/214901903-402c98c6-a44d-47ed-9b6e-6719dae8ac9f.png)
